man/f1tattoo.1
.\" f1tattoo is released under the GNU GENERAL PUBLIC LICENSE.
.TH "f1tattoo" "1" "10. April, 2009" "Gennady ShultZ Kozlov" "f1tattoo 0.7"
.br
.SH "NAME"
f1tattoo \- disc T@2 feature for Yamaha F1 CD-RW devices
.SH "SYNOPSIS"
.br
\fBf1tattoo [-l] [-h] \fr
.br
\fBf1tattoo -d DEVICE [-c] [-s] [--tattoo-test]\fR
.br
\fBf1tattoo -d DEVICE --tattoo-raw <file.raw>\fR
.br
\fBf1tattoo -d DEVICE --tattoo-png <file.png>\fR
.br
.SH "DESCRIPTION"
f1tattoo is the linux console tool to use the Yamaha disc T@2 feature. You won't get satisfying results on Phtalocyanin dye.
.SH "FEATURES"
You can disc T@2 with these drives:
.br
Yamaha F1 CD-RW, Yamaha F1DX

.SH "DEVICE"
can be an IDE, SCSI, SATA, USB or FireWire connected optical drive. Not all SATA controller support all Plextor features.

\fBLinux:\fR
.br
.I /dev/hdX:
IDE device
.br
.I /dev/scdX:
Linux 2.4: SATA, SCSI, USB device, or IDE device via ide-scsi emulation
.br
.I /dev/srX:
Linux 2.6: IDE device via new ATA layer, SCSI or USB device
.br

\fBOpenBSD/NetBSD:\fR
.br
.I /dev/rcdX
.br

\fBFreeBSD:\fR
.br
.I /dev/cd:
SCSI device
.br
.I /dev/acd:
ATA device
.br

\fBMacOS X:\fR
.br
.I /dev/disk:
.br

\fBwin32:\fR
.br
.I C:,D:,E:, ... X:,Y:,Z:
.br

.SH "OPTIONS"

\fB\-l, --scanbus\fR
.br
		scan busses for all available CD and DVD devices.

\fB\-h, --help\fR
.br
		help. show available options.

\fB\-c, --current\fR
.br
		show current drive settings of selected device

\fB\-s, --supported\fR
.br
		show supported features of selected device

\fB\--tattoo-test\fR
.br
		write DiscT@2 test image

\fB\--tattoo-raw tattoo.raw\fR
.br
		write DiscT@2 using raw data from tattoo.raw file

\fB\--tattoo-png tattoo.png\fR
.br
		write DiscT@2 using png image from tattoo.png file
		check if f1tattoo compiled with png support using --help option

\fB\-v, --verbose\fR
.br
		give verbose output
.SH "EXAMPLES"
\fBf1tattoo -l\fR
.br
		list all available CD- and DVD-devices\fR

\fBf1tattoo -d /dev/hdc \--tattoo-raw filename.raw\fR
.br
		writes DiscT@2 using filename.raw file on media in device /dev/hdc
\fR \fB
\fR \fB
\fR \fB
\fRplease report man page improvements to T.Maguin@web.de\fR


debian/changelog
qpxtool (0.7.2-4.1-ArtKarMOD) unstable; urgency=medium

  * support for BD-RE WH16NS58
  * other changes

 -- ArtKar <artkar@intercore.com.pl>  Mon, 15 Apr 2019 21:19:15 +0200

qpxtool (0.7.2-4.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Add patch 07-fix-wr-mode-declarations.patch:
    - Fix declaration of char attributes in wr_mode. (Closes: #856231)

 -- John Paul Adrian Glaubitz <glaubitz@physik.fu-berlin.de>  Mon, 27 Feb 2017 14:04:55 +0100

qpxtool (0.7.2-4) unstable; urgency=medium

  * Update debian/rules: delete option --parallel from dh command because it
    causes FTBFS on buildd hosts where parallel build is realy enabled.

 -- Boris Pek <tehnick@debian.org>  Sun, 13 Apr 2014 00:18:54 +0400

qpxtool (0.7.2-3) unstable; urgency=medium

  * Simplify patch 05-add-hardening-flags-in-compiler-options.patch.
  * Update debian/rules:
    - simplify rules after changes in related patch
    - add options --parallel and --list-missing to dh
  * Update debian/qpxtool.install.

 -- Boris Pek <tehnick@debian.org>  Sat, 12 Apr 2014 20:25:56 +0400

qpxtool (0.7.2-2) unstable; urgency=medium

  * Update debian/control:
    - udate using wrap-and-sort tool
    - delete now useless sections "Breaks: qpxtool (<< 0.7.1.002-1)"
    - add build dependency from hurd-dev on hurd-i386
  * Add patch 06-make-port-for-hurd-i386.patch:
    Make port for Debian GNU/Hurd.
  * Update debian/copyright.

 -- Boris Pek <tehnick@debian.org>  Sat, 12 Apr 2014 16:19:23 +0400

qpxtool (0.7.2-1) unstable; urgency=medium

  * New maintainer. (Closes: #629599)
  * Update to stable release 0.7.2.
  * Bump Standards-Version to 3.9.5 (was 3.9.3).
  * Update patches
    - add 01-do_not_use_lib64_dir.patch
    - rewrite 03-ftbfs_kfreebsd.patch
    - rewrite 02-desktop_file.patch
    - update 04-fix-build-with-libpng15.patch
    - delete 01-makefile.patch (it is unnecessary now)
  * Update debian/rules:
    - add necessary options to configure script for installing files
      in according with Debian Policy
    - add get-orig-source section
    - simplify rules
  * Fix lintian notes vcs-field-not-canonical.
  * Update years in debian/copyright.
  * Update debian/qpxtool.docs.
  * Add debian/qpxtool.doc-base.
  * Fix a lot of lintian notes hyphen-used-as-minus-sign:
    - add fixed debian/man/* files
    - add debian/qpxtool.manpages and debian/pxfw.manpages
    - update debian/qpxtool.install and debian/pxfw.install
  * Update debian/libqpx0.lintian-overrides.

 -- Boris Pek <tehnick@debian.org>  Wed, 26 Feb 2014 16:26:22 +0400

qpxtool (0.7.1.002-6) unstable; urgency=low

  * QA upload.
  * Bumped Standards-Version to 3.9.3 (was 3.9.2).
  * Bumped debhelper version to 9 (was 7.0.50~); updated debian/compat.
  * Added file debian/patches/05-add-hardening-flags-in-compiler-options.patch.
  * Updated debian/rules: added hardening flags in compiler options.
  * Changed build dependency from libpng12-dev to libpng-dev.
    Added file debian/patches/04-fix-build-with-libpng15.patch.
    Added small hack in debian/rules (see LIBPNG_VER variable).
    Build with libpng version 1.5.10 was tested successfully.
    (Closes: #662481, #648127)
  * Used [kfreebsd-any] instead of hardcoded list of kFreeBSD architectures
    [kfreebsd-i386 kfreebsd-amd64] in build dependency. (Closes: #634714)
  * File debian/copyright was updated in according to Copyright format 1.0.
  * Fixed such lintian warnings and notes:
    - out-of-date-copyright-format-uri
    - hardening-no-relro and hardening-no-fortify-functions

 -- Boris Pek <tehnick-8@mail.ru>  Tue, 26 Jun 2012 03:11:10 +0300

qpxtool (0.7.1.002-5) unstable; urgency=low

  * QA upload.
  * Bump Standards.
  * Orphaning this.

 -- Alessio Treglia <alessio@debian.org>  Tue, 21 Jun 2011 13:04:12 +0200

qpxtool (0.7.1.002-4) unstable; urgency=low

  * We need both Breaks and Replaces fields.

 -- Alessio Treglia <alessio@debian.org>  Wed, 06 Oct 2010 12:48:26 +0200

qpxtool (0.7.1.002-3) unstable; urgency=low

  * Link against libcam on kfreebsd to fix FTBFS.
  * debian/control:
    - Properly replace Conflicts/Replaces with Breaks fields.
    - Bump Standards.
  * Update debian/copyright.

 -- Alessio Treglia <alessio@debian.org>  Wed, 06 Oct 2010 09:09:15 +0200

qpxtool (0.7.1.002-2) unstable; urgency=low

  * Correct little typo in the descriptions, thanks to Davide Prina for
    reporting this (Closes: #582445).
  * Change my email address.

 -- Alessio Treglia <alessio@debian.org>  Sun, 06 Jun 2010 19:48:46 +0200

qpxtool (0.7.1.002-1) unstable; urgency=low

  * New maintainer (Closes: #543879).
  * New upstream release.
  * Move packaging to collab-maint's git area.
  * Split libraries:
    - libqpx0 (shared libs)
    - libqpx-dev (development files)
  * debian/control:
    - Drop libqt3-mt-dev, it's no longer needed.
    - Build-depend on qmake-qt4.
    - Bump Standards.
    - Set DM-Upload-Allowed to yes.
    - Build-depend on libqt4-dev.
    - Build-depend on libpng12-dev.
    - Set Conflicts/Replaces proper fields.
  * Switch to 3.0 (quilt) format.
  * Switch to debhelper 7.
  * Drop dpatch support:
    - Drop 02-typos.dpatch patch, no longer needed.
    - Drop 03-0.6.1.1.dpatch, no longer needed.
    - Drop 04-libata.dpatch patch, no longer needed.
    - 01-qtdir.dpatch -> 01-makefile.patch:
      + Drop QTDIR which seems to be not respected.
      + Set QMAKE4 properly.
      + Add DEP-3 compliant tags.
  * debian/patches/02-desktop_file.patch:
    - Fix desktop file as per spec.
  * debian/{libqpx0,qpxtool}.lintian-overrides:
    - "MultiMediaMasters & Machinary" is a manufacturer's name and not a
      spelling error.
  * debian/rules:
    - Cleanup.
    - Fix clean rule by passing the QMAKE4 envar.
  * Add watch file.

 -- Alessio Treglia <quadrispro@ubuntu.com>  Mon, 26 Apr 2010 09:36:21 +0200

qpxtool (0.6.1-5) unstable; urgency=low

  * Updating package to standards version 3.8.3.

console/f1tattoo/f1tattoo.cpp
/*
 * This file is part of the QPxTool project.
 * Copyright (C) 2005-2006,2009 Gennady "ShultZ" Kozlov <qpxtool@mail.ru>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * See the file "COPYING" for the exact licensing terms.
 */

#include <array>
#include <cstdio>
#include <cstdlib>

#include <qpx_mmc.h>
#include <yamaha_features.h>

#ifdef USE_LIBPNG
#include <png.h>
#endif

#define VERSION "0.7.1"

const uint32_t FL_HELP = 0x00000001;
const uint32_t FL_SCAN = 0x00000002;
const uint32_t FL_DEVICE = 0x00000004;
const uint32_t FL_VERBOSE = 0x00000008;
const uint32_t FL_CURRENT = 0x00000010;
const uint32_t FL_SUPPORTED = 0x00000020;
const uint32_t FL_TATTOO_TEST = 0x00000040;
const uint32_t FL_TATTOO_RAW = 0x00000080;
#ifdef USE_LIBPNG
const uint32_t FL_TATTOO_PNG = 0x00000100;
const uint32_t FL_TATTOO = FL_TATTOO_RAW | FL_TATTOO_PNG | FL_TATTOO_TEST;
#else
const uint32_t FL_TATTOO = FL_TATTOO_RAW | FL_TATTOO_TEST;
#endif

uint32_t flags = 0;

auto get_device_info(drive_info *drive) -> int {
    drive->ven_features = 0;
    drive->chk_features = 0;
    detect_capabilities(drive);
    //	detect_check_capabilities(drive);
    determine_disc_type(drive);
    if (isYamaha(drive) == 0) {
        printf("%s: drive not supported\n", drive->device);
        return 1;
    }
    //	if (!yamaha_check_amqr(drive)) drive->ven_features|=YMH_AMQR;
    //	if (!yamaha_check_forcespeed(drive))
    // drive->ven_features|=YMH_FORCESPEED;
    if (yamaha_f1_get_tattoo(drive) == 0) {
        drive->ven_features |= YMH_TATTOO;
    }

    if ((flags & FL_SUPPORTED) != 0U) {
        printf("\n** Supported features:\n");
        //		printf("AudioMaster Q.R.    : %s\n", drive->ven_features &
        // YMH_AMQR ? "YES" : "---"); 		printf("ForceSpeed          :
        //%s\n", drive->ven_features & YMH_FORCESPEED ? "YES" : "---");
        printf("DiscT@2             : %s\n",
               (drive->ven_features & YMH_TATTOO) != 0U ? "YES" : "---");
    }

    if ((flags & FL_CURRENT) != 0U) {
        printf("\n** Current drive settings:\n");
    }
    if ((flags & (FL_CURRENT | FL_TATTOO | FL_TATTOO_TEST)) != 0U &&
            (drive->ven_features & YMH_TATTOO) != 0U) {
        if ((drive->yamaha.tattoo_rows) != 0U) {
            printf("DiscT@2 info:\ninner: %d\nouter: %d\nimage: 3744x%d\n",
                   drive->yamaha.tattoo_i, drive->yamaha.tattoo_o,
                   drive->yamaha.tattoo_rows);
        } else {
            if ((drive->media.type & DISC_CD) != 0U) {
                printf("Can't write DiscT@2 on inserted disc!\n");
            } else {
                printf("No disc found! Can't get DiscT@2 info!\n");
            }
        }
    }
    return 0;
}

#ifdef USE_LIBPNG
static auto my_png_get_image_width(png_structp png_ptr, png_infop info_ptr)
-> int {
#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
    return png_get_image_width(png_ptr, info_ptr);
#else
    return info_ptr->width;
#endif
}

static auto my_png_get_image_height(png_structp png_ptr, png_infop info_ptr)
-> int {
#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
    return png_get_image_height(png_ptr, info_ptr);
#else
    return info_ptr->height;
#endif
}

static auto my_png_get_color_type(png_structp png_ptr, png_infop info_ptr)
-> png_byte {
#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
    return png_get_color_type(png_ptr, info_ptr);
#else
    return info_ptr->color_type;
#endif
}

static auto my_png_get_valid(png_structp png_ptr, png_infop info_ptr,
                             png_uint_32 flags) -> png_uint_32 {
#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
    return png_get_valid(png_ptr, info_ptr, PNG_INFO_PLTE);
#else
    return (info_ptr->valid & flags);
#endif
}

static auto my_png_get_bit_depth(png_structp png_ptr, png_infop info_ptr)
-> int {
#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
    return png_get_bit_depth(png_ptr, info_ptr);
#else
    return info_ptr->bit_depth;
#endif
}

static auto my_png_get_rowbytes(png_structp png_ptr, png_infop info_ptr)
-> int {
#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
    return png_get_rowbytes(png_ptr, info_ptr);
#else
    return info_ptr->rowbytes;
#endif
}

auto tattoo_read_png(unsigned char *buf, uint32_t rows, FILE *fp) -> bool {
    const int byte_size = 8;
    png_byte header[byte_size]; // 8 is the maximum size that can be checked
    png_structp png_ptr = nullptr;
    png_infop info_ptr = nullptr;
    uint32_t number_of_passes = 0;
    png_bytep png_row_pointer = nullptr;
    unsigned char *raw_row_pointer = nullptr;
    //	unsigned char *tp = NULL;

    //	int width;
    uint32_t row = 0;
    uint32_t col = 0;
    int c = 0;
    int32_t r = 0;
    int32_t g = 0;
    int32_t b = 0;
    int num_palette = 0;
#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
    png_colorp palette = nullptr;
#endif

    if (fread(header, 1, byte_size, fp) < byte_size) {
        printf("Error reading PNG header\n");
        fclose(fp);
        return true;
    }
    if (png_sig_cmp(header, 0, byte_size) != 0) {
        printf("File not recognized as a PNG\n");
        fclose(fp);
        return true;
    }
    png_ptr =
        png_create_read_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
    if (png_ptr == nullptr) {
        printf("png_create_read_struct failed!\n");
        fclose(fp);
        return true;
    }
    info_ptr = png_create_info_struct(png_ptr);
    if (info_ptr == nullptr) {
        printf("png_create_info_struct failed!\n");
        fclose(fp);
        return true;
    }
    if (setjmp(png_jmpbuf(png_ptr))) {
        printf("png_jmpbuf failed!\n");
        fclose(fp);
        return true;
    }

    png_init_io(png_ptr, fp);
    png_set_sig_bytes(png_ptr, byte_size);

    png_read_info(png_ptr, info_ptr);

    printf("Image size: %ld x %ld\n", my_png_get_image_width(png_ptr, info_ptr),
           my_png_get_image_height(png_ptr, info_ptr));

    if (my_png_get_image_width(png_ptr, info_ptr) != 3744U ||
            my_png_get_image_height(png_ptr, info_ptr) != rows) {
        printf("Image should be 3744 x %d", rows);
        return true;
    }

    //	width = info_ptr->width;
    //	height = info_ptr->height;
    //	bit_depth = info_ptr->bit_depth;

#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
    number_of_passes = png_set_interlace_handling(png_ptr);
#else
    number_of_passes = png_set_interlace_handling(png_ptr);
#endif
    png_read_update_info(png_ptr, info_ptr);

    printf("Color type: [%d] ", my_png_get_color_type(png_ptr, info_ptr));
    switch (my_png_get_color_type(png_ptr, info_ptr)) {
    case PNG_COLOR_TYPE_GRAY:
        printf("PNG_COLOR_TYPE_GRAY\n");
        break;
    case PNG_COLOR_TYPE_PALETTE:
        printf("PNG_COLOR_TYPE_PALETTE\n");
        if ((my_png_get_valid(png_ptr, info_ptr, PNG_INFO_PLTE)) == 0U) {
            printf("PNG color type is indexed, but no palette found!");
            goto err_read_png;
        }
        break;
    case PNG_COLOR_TYPE_RGB:
        printf("PNG_COLOR_TYPE_RGB\n");
        break;
    case PNG_COLOR_TYPE_RGB_ALPHA:
        printf("PNG_COLOR_TYPE_RGB_ALPHA\n");
        break;
    case PNG_COLOR_TYPE_GRAY_ALPHA:
        printf("PNG_COLOR_TYPE_GRAY_ALPHA\n");
        break;
    default:
        printf("unlnown PNG color type!\n");
        goto err_read_png;
    }
    printf("Bit depth : %d\n", my_png_get_bit_depth(png_ptr, info_ptr));
    if (my_png_get_bit_depth(png_ptr, info_ptr) != byte_size) {
        printf("Unsupported bit depth!\n");
        goto err_read_png;
    }

#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
    png_get_PLTE(png_ptr, info_ptr, &palette, &num_palette);
#else
    num_palette = info_ptr->num_palette;
#endif
    if (my_png_get_valid(png_ptr, info_ptr, PNG_INFO_PLTE) != 0U) {
        printf("Palette   : %d colors\n", num_palette);
    } else {
        printf("Palette   : NO\n");
    }
    printf("ROW bytes : %ld\n", my_png_get_rowbytes(png_ptr, info_ptr));

    raw_row_pointer = buf;
    png_row_pointer =
        static_cast<png_byte *>(malloc(my_png_get_rowbytes(png_ptr, info_ptr)));
    for (row = 0; row < rows; row++) {
        if (setjmp(png_jmpbuf(png_ptr))) {
            printf("png_jmpbuf failed!\n");
            goto err_read_png;
        }
        png_read_row(png_ptr, png_row_pointer, nullptr);
        if (my_png_get_image_width(png_ptr, info_ptr) < 3744U) {
            memset(raw_row_pointer, 0, 3744);
        }

        switch (my_png_get_color_type(png_ptr, info_ptr)) {
        case PNG_COLOR_TYPE_GRAY:
            for (col = 0; col < my_png_get_image_width(png_ptr, info_ptr); col++) {
                raw_row_pointer[col] = png_row_pointer[col] ^ 0xFF;
                //					memcpy(raw_row_pointer,
                // png_row_pointer, 3744);
            }
            break;
        case PNG_COLOR_TYPE_PALETTE:
            for (col = 0; col < my_png_get_image_width(png_ptr, info_ptr); col++) {
                c = png_row_pointer[col];
#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
                r = palette[c].red;
                g = palette[c].green;
                b = palette[c].blue;
#else
                r = info_ptr->palette[c].red;
                g = info_ptr->palette[c].green;
                b = info_ptr->palette[c].blue;
#endif
                c = (r * 11 + g * 16 + b * 5) / 32;
                raw_row_pointer[col] = c ^ 0xFF;
            }
#if PNG_LIBPNG_VER_MAJOR == 1 && PNG_LIBPNG_VER_MINOR >= 4
            png_set_PLTE(png_ptr, info_ptr, palette, num_palette);
#endif
            break;
        case PNG_COLOR_TYPE_RGB:
            for (col = 0; col < my_png_get_image_width(png_ptr, info_ptr); col++) {
                r = png_row_pointer[col * 3];
                g = png_row_pointer[col * 3 + 1];
                b = png_row_pointer[col * 3 + 2];
                c = (r * 11 + g * 16 + b * 5) / 32;
                raw_row_pointer[col] = c ^ 0xFF;
            }
            break;
        case PNG_COLOR_TYPE_RGB_ALPHA:
            for (col = 0; col < my_png_get_image_width(png_ptr, info_ptr); col++) {
                r = png_row_pointer[col * 4];
                g = png_row_pointer[col * 4 + 1];
                b = png_row_pointer[col * 4 + 2];
                c = (r * 11 + g * 16 + b * 5) / 32;
                raw_row_pointer[col] = c ^ 0xFF;
            }
            break;
        case PNG_COLOR_TYPE_GRAY_ALPHA:
            for (col = 0; col < my_png_get_image_width(png_ptr, info_ptr); col++) {
                raw_row_pointer[col] = png_row_pointer[col * 2] ^ 0xFF;
            }
            break;
        }
        raw_row_pointer += 3744;
    }

    //	if (tp) free(tp);
    if (png_row_pointer != nullptr) {
        free(png_row_pointer);
    }
    return true;

err_read_png:
    //	if (tp) free(tp);
    if (png_row_pointer != nullptr) {
        free(png_row_pointer);
    }
    return false;
}
#endif

void usage(char *bin) {
    fprintf(stderr, "\nusage: %s [-d device] [optinos]\n", bin);
#ifdef USE_LIBPNG
    printf("PNG support: YES\n");
#else
    printf("PNG support: NO\n");
#endif
    printf("\t-l, --scanbus                list drives (scan IDE/SCSI bus)\n");
    printf("\t-h, --help                   show help\n");
    printf("\t-c, --current                show current drive settings\n");
    printf("\t-s, --supported              show features supported by drive\n");
    printf("\t--tattoo-raw <tattoo_file>   burn selected RAW image as DiscT@2\n");
    printf("\t--tattoo-png <tattoo_file>   burn selected PNG image as DiscT@2\n");
#ifdef USE_LIBPNG
    printf("\t                             WARNING: f1tattoo compiled without "
           "libpng\n");
#endif
    printf("\t--tattoo-test                burn tattoo test image\n");
    printf("\t-v, --verbose                be verbose\n");
}

auto main(int argc, char *argv[]) -> int {
    int i = 0;
    char *device = nullptr;
    char *tattoofn = nullptr;
    unsigned char *tattoobuf = nullptr;
    drive_info *drive = nullptr;

    printf("**  DiscT@2 writer for Yamaha CRW-F1  v%s (c) 2005-2006,2009  "
           "Gennady \"ShultZ\" Kozlov  **\n",
           VERSION);

    for (i = 1; i < argc; i++) {
        //		printf("arg[%02d]: %s\n",i,argv[i]);
        if (strcmp(argv[i], "-d") == 0) {
            if (argc > (i + 1)) {
                i++;
                flags |= FL_DEVICE;
                device = argv[i];
            } else {
                printf("Option %s needs parameter\n", argv[i]);
                exit(1);
            }
        } else if (strcmp(argv[i], "-h") == 0) {
            flags |= FL_HELP;
        } else if (strcmp(argv[i], "--help") == 0) {
            flags |= FL_HELP;
        } else if (strcmp(argv[i], "-c") == 0) {
            flags |= FL_CURRENT;
        } else if (strcmp(argv[i], "--current") == 0) {
            flags |= FL_CURRENT;
        } else if (strcmp(argv[i], "-l") == 0) {
            flags |= FL_SCAN;
        } else if (strcmp(argv[i], "--scanbus") == 0) {
            flags |= FL_SCAN;
        } else if (strcmp(argv[i], "-s") == 0) {
            flags |= FL_SUPPORTED;
        } else if (strcmp(argv[i], "--supported") == 0) {
            flags |= FL_SUPPORTED;
        } else if (strcmp(argv[i], "-v") == 0) {
            flags |= FL_VERBOSE;
        } else if (strcmp(argv[i], "--verbose") == 0) {
            flags |= FL_VERBOSE;
        } else if (strcmp(argv[i], "--tattoo-test") == 0) {
            flags |= FL_TATTOO_TEST;
        } else if (strcmp(argv[i], "--tattoo-raw") == 0) {
            flags |= FL_TATTOO_RAW;
            if (argc > (i + 1)) {
                i++;
                tattoofn = argv[i];
            } else {
                printf("option %s needs parameter!\n", argv[i]);
                return 5;
            }
        } else if (strcmp(argv[i], "--tattoo-png") == 0) {
#ifdef USE_LIBPNG
            flags |= FL_TATTOO_PNG;
            if (argc > (i + 1)) {
                i++;
                tattoofn = argv[i];
            } else {
                printf("option %s needs parameter!\n", argv[i]);
                return 5;
            }
#else
            printf("Can't use PNG as input file: compiled without libpng");
#endif
        } else {
            printf("Illegal option: %s.\nUse -h for details\n", argv[i]);
            return 6;
        }
    }

    if ((flags & FL_HELP) != 0U) {
        usage(argv[0]);
        return 0;
    }
    if (flags == 0U) {
        usage(argv[0]);
        return 1;
    }
    if ((flags & FL_SCAN) != 0U) {
        int drvcnt = 0;
        drvcnt = scanbus(DEV_YAMAHA);
        if (drvcnt == 0) {
            printf("ERR: no drives found!\n");
        }
        return 2;
    }
    if ((flags & FL_DEVICE) == 0U) {
        printf("** ERR: no device selected\n");
        return 3;
    }

    //	printf("____________________________\n");
    printf("Device : %s\n", device);
    drive = new drive_info(device);
    if ((drive->err) != 0) {
        printf("%s: device open error!\n", argv[0]);
        delete drive;
        return 4;
    }
    inquiry(drive);
    //	convert_to_ID(drive);
    printf("Vendor : '%s'\n", drive->ven);
    printf("Model  : '%s'\n", drive->dev);
    printf("F/W    : '%s'\n", drive->fw);
    if ((flags & FL_VERBOSE) == 0U) {
        drive->silent++;
    }
    if (get_drive_serial_number(drive) != 0) {
        printf("Serial#: %s\n", drive->serial);
    }

    if (flags != 0U) {
        //	if (flags & FL_VERBOSE) {
        printf("\nf1tattoo flags : ");
        if ((flags & FL_DEVICE) != 0U) {
            printf(" DEVICE");
        }
        if ((flags & FL_HELP) != 0U) {
            printf(" HELP");
        }
        if ((flags & FL_CURRENT) != 0U) {
            printf(" CURRENT");
        }
        if ((flags & FL_SCAN) != 0U) {
            printf(" SCAN");
        }
        if ((flags & FL_VERBOSE) != 0U) {
            printf(" VERBOSE");
        }
        if ((flags & FL_SUPPORTED) != 0U) {
            printf(" SUPPORTED");
        }
        if ((flags & FL_TATTOO) != 0U) {
            printf(" TATTOO");
        }
        if ((flags & FL_TATTOO_TEST) != 0U) {
            printf(" TATTOO_TEST");
        }
        printf("\n\n");
    }
    get_device_info(drive);
    //	printf("____________________________\n");

    if ((flags & FL_TATTOO) != 0U) {
        FILE *tattoof = nullptr;
        bool fr = false;
        if ((drive->ven_features & YMH_TATTOO) == 0U) {
            printf("Selected device doesn't have DiscT@2 feature!\n");
            delete drive;
            return 1;
        }
        if ((flags & FL_TATTOO_TEST) != 0U) {
            printf("%s: writing T@2 test image...\n", device);
            yamaha_f1_do_tattoo(drive, nullptr, 0);
        } else {
            tattoof = fopen(tattoofn, "r");
            if (tattoof == nullptr) {
                printf("Can't open tattoo file: %s", tattoofn);
            } else {
                printf("Reading tattoo file...\n");
                tattoobuf = static_cast<unsigned char *>(
                                malloc(drive->yamaha.tattoo_rows * 3744));
#ifdef USE_LIBPNG
                if ((flags & FL_TATTOO_PNG) != 0U) {
                    fr = tattoo_read_png(tattoobuf, drive->yamaha.tattoo_rows, tattoof);
                } else {
#endif
                    memset(tattoobuf, 0, drive->yamaha.tattoo_rows * 3744);
                    fr = (fread((void *)tattoobuf, 3744, drive->yamaha.tattoo_rows,
                                tattoof) > 0);
#ifdef USE_LIBPNG
                }
#endif
                fclose(tattoof);
                if (fr) {
                    yamaha_f1_do_tattoo(drive, tattoobuf, static_cast<int>(fr) * 3744);
                } else {
                    printf("Error reading T@2 image!\n");
                }
                free(tattoobuf);
            }
        }
    }
    if ((flags & FL_VERBOSE) == 0U) {
        drive->silent--;
    }
    delete drive;
    return 0;
}
